Add usage of `EMParams`

johschnee opened the pull request:
## Changes

- Class `EMParams` is now used by the `ExpectationMaximizationModel` in `tools/_em_model.py`.

## Related issues

Closes #1026, #1025.

johschnee commented:
@WeilerP Currently, I have the following questions about the future design of the API:
1. Should all the fitted parameters still be written in the AnnData object by default when calling `ExpectationMaximiziationModel.fit()`? Or should the `EMParams` class become the new way of handling the inferred parameters and only be written to the AnnData object when calling `ExpectationMaximiziationModel.export_results_adata()` (not yet implemented)?
2. My current implementation relies on the methods `_read_pars()` and `_write_pars()` in `tools/_em_model_core.py`. Should these methods be kept in the implementation at all or do you aim to remove them with this restructuring?
3. In the old method `recover_dynamics()` the parameter copy is false by default. In the new implementation `ExpectationMaximizationModel.fit()` it is true by default. Is this on purpose?

WeilerP commented on scvelo/tools/_em_model.py:
```diff
@@ -126,6 +125,10 @@ def __init__(
         self._backend = backend
 
+    def _initialize_state_dict(self, adata, pars_names=None, key="fit"):
+        pars = _read_pars(adata, pars_names, key)
```
Can you please refactor the code s.t. we do not call `_read_pars`? We'll remove `_em_model_core.py` and `_em_model_utils.py` in the future.

johschnee replied:
Of course, this refactoring is done.

WeilerP commented on scvelo/tools/_em_model.py:
```diff
@@ -274,7 +264,7 @@ def fit(
         }
 
-        _write_pars(adata, _pars, add_key=self._fit_key)
+        _write_pars(adata, list(asdict(sd).values()), add_key=self._fit_key)
```
Same as for `_read_pars`. Also, let's use the `export_results_adata` function for writing parameters and not have it part of the `fit` method.

johschnee replied:
I used `export_results_adata`, but I noticed that `fit` does not work without writing the parameters during fitting (or at least it would be a larger refactoring, which I'd want to do in a separate commit). The parameters written to the adata object are used by the `align_dynamics` method, which is called during `fit` after writing.

`align_dynamics` also modifies the written values. Hence, `export_results_adata` currently gives different results than there are returned by `fit`.

For refactoring, I'd suggest to modify `align_dynamics` as well such that is uses `EMParams` and copy it to `_em_model.py` (currently the method is in `_em_model_core.py`). Is this approach okay or would you suggest something different?

johschnee commented:
Yes, I was working on it actually today. I added the implementation of the abstract methods to this PR, because `export_results_adata` is used instead of `_write_pars`.

WeilerP commented on scvelo/tools/_em_model.py:
```diff
@@ -158,15 +168,17 @@ def _prepare_genes(self):
     def state_dict(self):
         """Return the state of the model."""
-        raise NotImplementedError
+        return asdict(self._state_dict)
```
Can you please undo this implementation? This PR should only concern switching from the current implementation to using `EMParams`. The implementation of these functions is covered in #1025.

WeilerP commented on scvelo/tools/_em_model.py:
```diff
@@ -126,6 +119,23 @@ def __init__(
         self._n_jobs = get_n_jobs(n_jobs=n_jobs)
         self._backend = backend
 
+    def _initialize_state_dict(
```
Why do we need the argument `parameters` and `key` if we never pass them to the function?

WeilerP commented on scvelo/tools/_em_model.py:
```diff
@@ -158,15 +168,17 @@ def _prepare_genes(self):
-    def export_results_adata(self):
-        """Export the results to the AnnData object."""
-        raise NotImplementedError
+    def export_results_adata(self, copy: bool = True, add_key: str = "fit"):
```
See my comment for the `state_dict` function: These functions should be implemented in a separate PR for maximal readability, etc.

WeilerP commented on scvelo/tools/_em_model.py:
```diff
@@ -239,42 +249,30 @@ def fit(
             L.append(dm.loss)
 
-        _pars = [
-            alpha,
-            beta,
-            gamma,
-            t_,
-            scaling,
-            std_u,
-            std_s,
-            likelihood,
-            u0,
-            s0,
-            pval,
-            steady_u,
-            steady_s,
-            varx,
-        ]
-
-        adata = self._adata.copy() if copy else self._adata
+        adata = self.export_results_adata(copy=copy, add_key=self._fit_key)
 
         adata.uns["recover_dynamics"] = {
             "fit_connected_states": fit_connected_states,
             "fit_basal_transcription": self._fit_basal_transcription,
             "use_raw": use_raw,
         }
 
-        _write_pars(adata, _pars, add_key=self._fit_key)
```
Similar to the implemented classes, can you please revert these changes and open a separate PR for them? Or am I missing something, and all of it is needed when switching to using `EMParams`?

WeilerP reviewed (commented):
@johschnee, I left another few comments. Mainly two things:
1. I updated/refactored the `_initialize_state_dict` function. Can you please double check that it is still working?
2. Can you please und all changes irrelevant to this PR (_i.e._, refactoring the code to rely on `EMParams`)? These changes (_e.g._, function implementations) should be there own PR accompanied by an issue.

WeilerP commented:
FYI, @johschnee, the _ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (4,) + inhomogeneous part._ error is caused by a new Numpy version (see #1058).

johschnee commented:
Hi @WeilerP,
this branch now contains the changes we discussed. In particular, I changed the following:
- Added metadata to the attributes of `EMParams` and read/write methods: The metadata allows a different handling of vector and matrix parameters without hardcoding them in the read/write methods. I decided to add the read/write methods to EMParams because they rely a lot on the attributes of EMParams.
- In the method `fit()`, I used the class EMParams. Since the AnnData object should not be modified during fit(), I defined some private attributes to store results and additional settings: `_loss`, `_fit_connected_states`, and `_use_raw`. They are written to an AnnData object when `export_results_adata()` is called.
- As we discussed, I copied `_align_dynamics()` to the class `ExpectationMaximizationModel` and adapted it slightly to use the EMParams. Now, `ExpectationMaximizationModel` only uses the class `DynamicsRecovery` from `_em_model_core`, but no other methods.

Overall, 
```
model = ExpectationMaximizationModel(adata)
model.fit()
adata_return = model.export_results_adata()
```
and
```
scv.tl.recover_dynamics(adata)
```
should lead to the same results. I tested it with simulated data and the pancreas dataset.

johschnee commented:
Hi @WeilerP,
I changed the behavior of `export_to_adata()` such that arrays containing only nan values are not written anymore.
I also compared the results of the old and the new API for the simulated data and the pancreas dataset using numpy. As expected, they were identical.
Hence, it would be great if you could review this PR.

WeilerP reviewed (approved):
Thanks, @johschnee!
